### DISCUSSION TITLE
I would like access to some canvas variable for meta data

### DISCUSSION THREAD

**Discussion by sanderjson** [darklang/dark#4345] [Suggestions]
- id/name of all handlers with uri
- recent trace data
- function names, variables, and types
- handler call frequency

**Comment by pbiggar** [comment]
Thanks! Would be interested to have a bit more detail about what the metadata would be used for, in each case?

**Reply by sanderjson** [reply]
At the moment I am most interested in generating documentation for my canvas from meta data. I have a front-end tool already prepared which might be able to consume the data and generate documentation as a static asset. This might include:

- list of all handlers with uri and method (http)
- list of all functions with parameters and parameter types
- list of all datastores with key name and value type
- generally most information that is available in the Dark UI 

The purpose is to help plan development, track my own progress, possibly train others. Furthermore, I can build in my own analytics per handler but I wonder how much you already have access to. For instance, at some point Dark users might be interested in the number of requests in a time period (feature validation).  

Personally, I like the idea of auto-generating docs from this hypothetical canvas variable, and i don't believe it is far fetched to consider this service might be useful for Dark startups.

Also, as my canvas grows it is becoming more of a chore just to find and reference information. 

And finally, as a potential customer I do worry about data loss. If I could some how store my own canvas state then I would feel more comfortable relying on Dark for essential business infrastructure. I believe this hypothetical "real-time documentation snapshot" might be a step in the right direction.

**Reply by StachuDotNet** [reply]
> Personally, I like the idea of auto-generating docs from this hypothetical canvas variable, and i don't believe it is far fetched to consider this service might be useful for Dark startups.

Are you looking to generate docs for a developer of your project, or docs for the end user?
I think Dark including a built-in "documentation" feature would make sense, and would support the end user docs.
As for developer "docs," that should be improved by eventual canvas organization changes - autolayout, visual grouping, etc. Plenty to think through there, still.

**Comment by sanderjson** [comment]
I am essentially looking for both, because I represent both an end user and developer for my own project.

For me, I know believe there are three lines of thinking here:

1. Writing own app Documentation as a best practice
2. Access to canvas meta data for "in-house" analytics
3. Access to both to manage _building_ a large application.

At the moment I am most interested in the third. As my application is growing in size (canvas surface area) and complexity I am finding it increasingly difficult to build upon _and improve_.

It is likely that I am missing some existing best practices (education). This may include Dark UI "tips and tricks", functional programming best practices, and API development strategies. I am very interested in these topics and have UXD and IA (information architecture) experience.

What is the best way for similarly minded Dark users to facilitate this conversation and learn or teach each other?
